Add Typescript annotations to TableSelectRow

jpsorensen opened the pull request:
Closes #12531 

Convert existing propType definitions to Typescript annotations on the TableSelectRow component. This is part of a broader effort to add Typescript annotations to components, tracked in https://github.com/carbon-design-system/carbon/issues/12513.

#### Changelog

**Changed**

- TableSelectRow.js has been converted to tsx

#### Testing / Reviewing

- In the packages/react directory, run yarn build, yarn storybook. In the storybook, navigate to DataTable>Selection and confirm that the component's behavior remains unchanged.
- Past usages of TableSelectRow at the bottom of TableSelectRow.tsx. Confirm that they compile as expected. Modify inputs so that they have the wrong types and confirm that it does not compile. Remove a required parameter and confirm that it does not compile. (I used the examples from the TableSelectRow unit tests).

jpsorensen committed:
refactor(NumberInput): add typescript annotations

Convert existing propType definitions to Typescript annotations on the
NumberInput component. This is part of a broader effort to add
Typescript annotations to components, tracked in #12513. Closes #12548.

Type annotation changes only; no breaking feature changes.

jpsorensen committed:
Merge branch 'main' into refactor/12531-table-select-row-typescript

jpsorensen committed:
Merge branch 'main' into refactor/12531-table-select-row-typescript

jpsorensen committed:
fix(typescript): disable no-non-null when safe

Disable eslint no-no-null-assertion rule on a line where it is known to be safe
(because it will ultimately pass in a key-value pair of undefined: undefined,
which will have no effect on the class names produced).

jpsorensen commented on packages/react/src/components/DataTable/TableSelectRow.tsx:
```diff
@@ -12,6 +12,57 @@ import InlineCheckbox from '../InlineCheckbox';
+  ariaLabel: string;
+
+  /**
+   * Specify whether all items are selected, or not
```
I'm guessing that this was accidentally copied over from TableSelectAll. I decided to leave it as-is (leave it in the PropTypes docs, and to have the Typescript docs mirror the PropTypes docs) in this PR so as not to get a separate change mixed up in this Typescript conversion. Just let me know if you'd prefer that I fix this as part of this PR!

francinelucca replied:
nice catch! I do believe that's a mistake, I think it'll be fine to fix here as to not carry over more mistakes, we can rename here and in the proptypes to something like "Specify whether row is selected, or not" if you don't mind adding it 🙏🏻

francinelucca commented on packages/react/src/components/DataTable/TableSelectRow.tsx:
```diff
@@ -34,7 +85,8 @@ const TableSelectRow = ({
   const tableSelectRowClasses = classNames(`${prefix}--table-column-checkbox`, {
-    [className]: className,
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    [className!]: className,
```
think this way you can forego having to disable the eslint rule
```suggestion
    ...(className && {[className]: true}),
```

jpsorensen replied:
Done, thank you!

francinelucca reviewed (approved):
Thanks for getting this PR up! LGTM, just two minor details

jpsorensen committed:
fix(typescript): fix docs and reinstante no-non-null

Fix doc for checked, which was incorrect in the propTypes.
Instead of disabling no-non-null and using an assertion,
check whether className is defined before adding it.

kodiakhq[bot] committed:
Merge branch 'main' into refactor/12531-table-select-row-typescript
